_adr/88/index.adoc

```diff
@@ -1,48 +1,71 @@
 ---
 
-// Top style tips:
-// * Use one sentence per line
-// * No unexpanded acronyms
-// * No undefined jargon
-
-// No need for a title heading, it's added by the template
-
 ## Context and Problem Statement
-// What is the background against which this decision is being taken?
+The purpose of this ADR is to define a standardized approach for securing admin API in a fleet manager
 
 ## Goals
-// Bulleted list of outcomes that this ADR, if accepted, should help achieve
+* Provide standardized way to secure admin API endpoints for fleet managers
 
 ## Non-goals
-// Bulleted list of outcomes that this ADR is not trying to achieve.
 
 ## Current situation
-// Where are we now?
 
 ## Proposal
-// What is the decision being proposed
+
+image::adr-88-admin-api-architecture-diagram.png[architecture diagram]
+
+A Keycloak realm is used to authorize access to admin API endpoint(s). 
+Such realm contains roles that allow access to endpoints to the privileged API users.
+Recommended approach would be to create a set of realm roles per environment and name them in the way that indicates the permissions they grant as follows:
+
+* <unique-fleet-manager-prefix>-admin-read-stage
+* <unique-fleet-manager-prefix>-admin-write-stage
+* <unique-fleet-manager-prefix>-admin-full-stage
+
+* <unique-fleet-manager-prefix>-admin-read-prod
+* <unique-fleet-manager-prefix>-admin-write-prod
+* <unique-fleet-manager-prefix>-admin-full-prod
+
+These roles map to the following permissions for the environment included as the role postfix (e.g. `prod`):
+
+* \*-read-* - list permissions
+* \*-write-* - list/ update permissions
+* \*-full-* - full permissions (list/ update/ delete)
+
+Example full role for kas-fleet-manager in production environment is: `kas-fleet-manager-admin-full-prod`.
+
+Once Keycloak instance is set up, a privileged user uses their credentials to get a JWT token from Keycloak.
+The JWT token is sent to a admin API in a HTTP request.
+Admin API authorization middleware verifies token signature with Public JWKS Certificate and roles included in the JWT token are checked against roles specified in the fleet manager.
+Matching role for given admin endpoint grants access to the action associated with that endpoint.
+Failure to match the required role results in restricting access to the endpoint.
+
+image::adr-88-authentication-flow.png[architecture diagram]
+
+## Suggested implementation for Red Hat Internal Keycloak
+
+Instead of using dedicated Keycloak instance/ realm, roles allowing to access admin API endpoints are represented by membership in a Red Hat Rover Group(s).
 
 ### Threat model
-// Provide a link to the relevant threat model. 
-// You must either update an existing threat model(s) to cover the changes made by this ADR, or add a new threat model.
+Admin API endpoints are protected by realm roles, hence only users that are assigned these roles can access the endpoints.
+
+No additional mitigation is required.
```

Comment:
I don't think the sentences added meet the requirements in the comments removed.

Reply:
100% agree that the concrete kas-fleet-manager threat model will need to be updated. I am reaching out to product security for more details on this process.

For this ADR, I understood the intent was not to link to Red Hat specific implementations since this ADR is generic, and there is no generic threat model that can be modified/attached, what is your opinion? @emmanuelbernard @tombentley

Reply:
@pawelpaszki that sounds reasonable. Can we add a sentence something like "This is a generic ADR so there is no threat model to update. The application of this ADR to specific services will require updating the threat model for those services"